### ISSUE TITLE
bug(offpolicy): 多卡 optimizer CUDA Graph 捕获 NCCL all-reduce 时死锁

### ISSUE THREAD

**Issue by TATP-233** [Motphys/UniLab#978]
## 问题摘要

在 off-policy 单机多卡数据并行中，如果把每次 optimizer update 的 flat-gradient `torch.distributed.all_reduce` 放进 actor / critic CUDA Graph，两个 rank 会在第一次 graph capture 阶段永久等待。相同模型、batch 和 NCCL process group 使用 eager optimizer update 时可以稳定完成训练；单卡 optimizer CUDA Graph 也正常。

当前 mitigation 已在 `7f31dcbb` 落地：多卡配置请求 actor/critic optimizer graph capture 时自动切换到 eager update，runtime manifest 记录 `dp_sync.cuda_graph_optimizer_capture=eager_fallback`。本 issue 跟踪 NCCL capture 的根因和后续是否能安全恢复真正的多卡 CUDA Graph，而不是移除该 fallback。

关联：#964、Discussions #975/#976/#977。

## 实测环境

- GPU：2 × NVIDIA RTX 6000D 85GB
- Driver：595.84
- PyTorch：2.7.0+cu128
- CUDA：12.8
- NCCL：2.26.2
- topology：每卡一个 learner + 一个 collector，rank-local replay
- rendezvous：FileStore
- NCCL transport：TCP loopback；`NCCL_P2P_DISABLE=1`、`NCCL_SHM_DISABLE=1`

## 复现方式

在 `DpParameterSync.allreduce_gradients()` 保持 flat-gradient buffer 生命周期覆盖 graph replay，并在 graph capture candidate 中按以下顺序插入同步：

```text
optimizer.zero_grad(set_to_none=True)
loss.backward()
flat gradients -> all_reduce(SUM) -> divide world_size -> copy back
optimizer.step()
```

然后运行：

```bash
uv run scripts/train_offpolicy.py \
  algo=flashsac \
  task=flashsac/g1_walk_flat/mujoco \
  training.devices=[0,1] \
  training.no_play=true \
  training.use_amp=false \
  algo.num_envs=4096 \
  algo.replay_buffer_n=256 \
  algo.updates_per_step=2 \
  algo.learning_starts=1 \
  algo.max_iterations=3 \
  algo.algo_params.use_cuda_graph_critic=true \
  algo.algo_params.use_cuda_graph_actor=true \
  algo.algo_params.use_cuda_graph_critic_packed_staging=true \
  algo.algo_params.use_cuda_graph_actor_packed_staging=true
```

## 实际结果

- 两个 rank 均完成 process-group 初始化、startup model broadcast、collector 启动和 replay warmup。
- 第一次 optimizer graph capture 后训练不再前进；持续超过 2 分钟无异常、无 timeout、无 Xid。
- 两个进程各保持约 3.8 GiB GPU 显存，需要外部终止。
- 把 graph 选项关闭后，相同命令立即通过；FastSAC 和 FlashSAC 的 eager per-optimizer gradient all-reduce 均已通过真实双卡 smoke。

## 预期结果

若当前 NCCL/PyTorch/transport 组合支持 capture，第一次 capture 和后续 replay 应在所有 rank 上按相同 collective 顺序完成。若组合不受支持，应在 capture 前 fail-fast 或自动 fallback，不能永久挂起。

## 调查边界

本 issue 只回答一个主要问题：确认该环境下 NCCL collective capture 的支持条件和死锁根因，并给出可验证的安全结论。不在本 issue 中改变 rank-local replay、runner lifecycle 或梯度同步语义。

重点检查：

- TCP-loopback NCCL transport 是否支持 CUDA Graph capture；
- `async_op=False` / Work wait 与 capture stream 的交互；
- capture mode、rank 间首次 capture 顺序和 communicator warmup 要求；
- flat-gradient buffer address/lifetime 与 graph memory pool；
- PyTorch 2.7.0 / NCCL 2.26.2 是否存在已知限制或需额外环境配置。

## 验收标准

- 提供最小双 rank NCCL capture reproducer，带有限 timeout，能够稳定复现或证明已解决；
- 明确记录支持矩阵及必要配置；
- 若支持：FastSAC 与 FlashSAC 各完成一次开启 actor/critic graph 的双卡 smoke，并验证每个实际 optimizer update 都执行梯度 collective；
- 若不支持：保留 eager fallback，并补充明确诊断与回归测试，确保任何配置都不会进入永久等待。

**Comment by TATP-233** [comment]
## 验收结论（2026-08-16）

本 issue 的有限超时复现、支持条件、双算法 CUDA Graph 真机验证和回归测试均已完成。

- **有限 timeout reproducer**：`benchmark/rl/reproduce_nccl_cuda_graph_capture.py` 使用父进程 20 秒 timeout 和 2-rank NCCL。目标环境中，无 eager collective warmup 时 capture 稳定报错；完成 eager warmup 后，`async_op` 同步/异步与 default/side stream 四种组合均可 capture/replay，不再永久等待。
- **支持条件**：SAC 用户文档已记录实测 PyTorch/CUDA/NCCL/RTX 6000D 环境，以及 capture 前 eager collective warmup、跨 replay 稳定 flat-gradient buffer、graph teardown；目标平台默认禁用 NCCL P2P/SHM，用户显式环境变量仍优先。
- **FastSAC 双卡 smoke**：actor/critic optimizer graph 与 packed staging 均实际启用，2-rank 运行 4 iterations 后自然退出；runtime manifest 为 `gradient_mean_per_optimizer_step` / `enabled_after_collective_warmup`，每轮 `train/dp_gradient_sync_calls=5`，与该轮实际 optimizer updates 一致。
- **FlashSAC 双卡 smoke**：同样启用 actor/critic optimizer graph 与 packed staging，2-rank 运行 4 iterations 后自然退出；manifest 同上，每轮 `train/dp_gradient_sync_calls=4`，与实际 optimizer updates 一致。
- **回归验证**：最终集成树已通过 `make test-all`：1657 passed、24 skipped、1 xfailed，coverage 72%，benchmark import smoke 通过。
- **落地记录**：修复由 #979 合入 `feat/issue-964-offpolicy-multi-gpu-dp`，#981 包含最终 PPO 集成及退出清理；后续随 #964 umbrella PR 进入 `main`。

因此 #978 的全部验收标准已经满足，按 completed 关闭。
